### PR TITLE
Pass mouse input to IME popups

### PR DESCRIPTION
**Pull request by MikeWalrus**
#### Describe your PR, what does it fix/add?
This pr allows user to interact with IME popups with their mice.

#### Is there anything you want to mention? (unchecked code, possible bugs, found problems, breaking compatibility, etc.)
It's my first time looking into the code base, so I might miss something. :D

#### Is it ready for merging, or does it need work?
Mouse input works with fcitx5 and Alacritty, but might need more testing.

https://github.com/hyprwm/Hyprland/assets/70618504/7f2d16dd-9e16-4350-84bc-04b3e8bc0e47

### CODE REVIEW

File: src/Compositor.cpp [hunk 1]
```diff
@@ -1109,6 +1109,20 @@ wlr_surface* CCompositor::vectorToLayerSurface(const Vector2D& pos, std::vector<
     return nullptr;
 }
 
+SIMEPopup* CCompositor::vectorToIMEPopup(const Vector2D& pos, std::list<SIMEPopup>* popups) {
```
Comment (vaxerski) [inline src/Compositor.cpp:1112]: why not a `const &`?

File: src/Compositor.cpp [hunk 2]
```diff
@@ -1109,6 +1109,20 @@ wlr_surface* CCompositor::vectorToLayerSurface(const Vector2D& pos, std::vector<
...
+    for (auto& popup : *popups) {
+        auto surface = popup.pSurface->surface;
+        auto realX   = popup.realX;
```
Comment (vaxerski) [inline src/Compositor.cpp:1115]: what for? `CBox box{popup.realX, popup.realY ...`

File: src/managers/input/InputManager.cpp [hunk 3]
```diff
@@ -240,6 +240,15 @@ void CInputManager::mouseMoveUnified(uint32_t time, bool refocus) {
     if (!foundSurface)
         foundSurface = g_pCompositor->vectorToLayerSurface(mouseCoords, &PMONITOR->m_aLayerSurfaceLayers[ZWLR_LAYER_SHELL_V1_LAYER_TOP], &surfaceCoords, &pFoundLayerSurface);
 
+    // then IME popups
+    if (!foundSurface) {
```
Comment (vaxerski) [inline src/managers/input/InputManager.cpp:244]: according to how it's rendered, popups are above TOP layers, and below OVERLAY ones. This should be moved a bit up.

File: src/Compositor.cpp (hunk 1 above)
Comment (MikeWalrus) [reply src/Compositor.cpp:1112]: You mean
```suggestion
SIMEPopup* CCompositor::vectorToIMEPopup(const Vector2D& pos, const std::list<SIMEPopup>& popups) {
```
and using `const_cast` to return the pointer? I don't see much use of `const_cast` in the code base.

Comment (vaxerski) [reply src/Compositor.cpp:1112]: I don't think you need to use const cast?

oh, maybe, cuz it's a list.

Then just & without const. My bad

Comment (DRAGONTOS) [reply src/Compositor.cpp:1112]: No do not do that it will fail if you compile it with changing it to &

Comment (vaxerski) [reply src/Compositor.cpp:1112]: ?

**Comment by MikeWalrus** [comment]
I believe the clean up logic of IME has some issue that will cause crashes. Not sure if it's out of the scope of this pr.

**Comment by vaxerski** [comment]
yea, it's tracked @ #2146

It's low prio to me, because you shouldn't restart IMEs anyways.

I use fcitx5 on my main machine and never had it crash.

Might look into it though when I have some time and dont forget

**Comment by vaxerski** [comment]
yea, it's tracked @ #2146

It's low prio to me, because you shouldn't restart IMEs anyways.

I use fcitx5 on my main machine and never had it crash.

Might look into it though when I have some time and dont forget

File: src/Compositor.cpp [hunk 4]
```diff
@@ -1109,6 +1109,21 @@ wlr_surface* CCompositor::vectorToLayerSurface(const Vector2D& pos, std::vector<
     return nullptr;
 }
 
+SIMEPopup* CCompositor::vectorToIMEPopup(const Vector2D& pos, std::list<SIMEPopup>& popups) {
+    for (auto& popup : popups) {
+        auto surface = popup.pSurface->surface;
+        CBox box{
```
Comment (vaxerski) [inline src/Compositor.cpp:1115]: did clang-format allow this? or did you forget to run it?

Comment (MikeWalrus) [reply src/Compositor.cpp:1115]: This was formatted by clang-format.

Comment (vaxerski) [reply src/Compositor.cpp:1115]: lol ok thats odd.

**Review by vaxerski — APPROVED** [review]
lgtm, thanks!

**Comment by MikeWalrus** [comment]
Thank you! @vaxerski

**Pull request merged**